### PR TITLE
Implemented check for slices for TPC QC

### PR DESCRIPTION
**Pull request by HorstMa**

### CODE REVIEW

File: Modules/TPC/src/CheckOfSlices.cxx [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+  const std::vector<double> vErr(yErrors, yErrors + NBins); // use all points
+
+  // now we have the mean and the stddev. Now check if all points are within the margins (default 3,6 sigma margins)
+
+  if (mCheckChoice == mCheckChoiceMean || mCheckChoice == mCheckChoiceBoth) {
+
+    // const std::vector<double> v(yValues + NBins - 1 - pointNumberForMean, yValues + NBins - 1);
+
+    const double sum = std::accumulate(v.begin(), v.end(), 0.0);
+    const double meanFull = sum / NBins;
```
Comment (wiechula) [inline Modules/TPC/src/CheckOfSlices.cxx:146]: Check for division by 0?

File: Modules/TPC/src/CheckOfSlices.cxx [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+    for (const double& yvalue : v) {
+      int i = &yvalue - &v[0];
+      double yError = vErr[i];
```
Comment (wiechula) [inline Modules/TPC/src/CheckOfSlices.cxx:150]: More readable?
```suggestion
    for (size_t i = 0; i < v.size(); ++i) {
      const auto yValue = v[i]
      const auto yError = vErr[i];
```

File: Modules/TPC/src/CheckOfSlices.cxx [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+      //}
+    }
+  } // if (mCheckChoice == mCheckChoiceMean || mCheckChoice == mCheckChoiceBoth)
+  result = resultMean;
+  //########################## ExpectedPhysicsValue #######################################
+  if (mCheckChoice == mCheckChoiceExpectedPhysicsValue || mCheckChoice == mCheckChoiceBoth) {
+
+    for (const double& yvalue : v) {
+      int i = &yvalue - &v[0];
+      double yError = vErr[i];
```
Comment (wiechula) [inline Modules/TPC/src/CheckOfSlices.cxx:176]: see above

File: Modules/TPC/src/CheckOfSlices.cxx [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    ILOG(Error, Support) << h->GetName() << ENDM;
+  }
+  const int NBins = h->GetN();
+
+  const double* yValues = h->GetY();
+  const double* yErrors = h->GetEY();
+  const std::vector<double> v(yValues, yValues + NBins); // use all points
+  const std::vector<double> vErr(yErrors, yErrors + NBins);
+  const double sum = std::accumulate(v.begin(), v.end(), 0.0);
+  const double meanFull = sum / NBins;
```
Comment (wiechula) [inline Modules/TPC/src/CheckOfSlices.cxx:233]: Catch division by 0?

File: Modules/TPC/src/CheckOfSlices.cxx [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    msg->AddText(Form("Mean: %f", meanFull));
+  }
+  const double xMin = h->GetXaxis()->GetXmin();
+  const double xMax = h->GetXaxis()->GetXmax();
+  TLine* lineExpectedValue = new TLine(xMin, mExpectedPhysicsValue, xMax, mExpectedPhysicsValue);
+  lineExpectedValue->SetLineColor(kGreen);
+  lineExpectedValue->SetLineWidth(2);
+  // mean Line
+
+  ILOG(Error, Support) << meanFull << ENDM;
```
Comment (wiechula) [inline Modules/TPC/src/CheckOfSlices.cxx:286]: Needed?
```suggestion
```

**Review by wiechula — CHANGES_REQUESTED** [review]

File: Modules/TPC/include/TPC/CheckOfSlices.h [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  ClassDefOverride(CheckOfSlices, 2);
+  std::string mCheckChoice;
+  float mExpectedPhysicsValue;
+  float mNSigmaExpectedPhysicsValue;
+  float mNSigmaMediumExpectedPhysicsValue;
+  float mNSigmaBadExpectedPhysicsValue;
+  float mNSigmaMean;
+  float mNSigmaMediumMean;
+  float mNSigmaBadMean;
+  float meanFull;
```
[the remark this answers is not in the document]
Comment (HorstMa) [reply Modules/TPC/include/TPC/CheckOfSlices.h]: not really required. the default values are set in the .cxx

**Review by wiechula — APPROVED** [review]

**Comment by knopers8** [comment]
Hi @wiechula , the failure on mac looks legit:
```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DBOOST_ALL_NO_LIB -DBOOST_CONTAINER_DYN_LINK -DBOOST_ERROR_CODE_HEADER_ONLY -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_REGEX_DYN_LINK -DBOOST_SERIALIZATION_DYN_LINK -DDPL_ENABLE_BACKTRACE -DFAIRMQ_HAS_STD_FILESYSTEM=1 -DFMT_LOCALE -DFMT_SHARED -DO2QcTPC_EXPORTS -I/System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/SOURCES/QualityControl/1209/0/Modules/TPC/include -I/System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/SOURCES/QualityControl/1209/0/Modules/TPC/src -I/System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/SOURCES/QualityControl/1209/0/Modules/TPC -I/System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/SOURCES/QualityControl/1209/0/Framework/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/O2/1209-local1/include/GPU -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/boost/v1.75.0-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/FairLogger/v1.11.0-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/fmt/8.1.1-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/FairMQ/v1.4.51-local2/include/fairmq -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/FairMQ/v1.4.51-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/ROOT/v6-24-06-patches-alice1-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/Common-O2/v1.6.0-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/libInfoLogger/v2.4.4-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/Monitoring/v3.12.0-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/Configuration/v2.6.3-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/Control-OCCPlugin/v0.44.1-local1/include/Occ -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/grpc/v1.45.1-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/protobuf/v3.14.0-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/c-ares/1.18.1-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/re2/2019-09-01-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/abseil/20211102.0-local2/include -isystem /usr/local/opt/openssl@1.1/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/O2/1209-local1/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/curl/7.70.0-local1/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/FairRoot/v18.4.7-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/arrow/v5.0.0-alice3-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/Vc/1.4.1-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/ms_gsl/4.0.0-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/libjalienO2/0.1.3-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/RapidJSON/v1.1.0-alice2-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/libuv/v1.40.0-local2/include -isystem /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/osx_x86-64/TBB/v2021.5.0-local2/include -fPIC -O2 -std=c++17 -Werror -Wno-error=deprecated-declarations -Wno-error=unused-function -O2 -g -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk -mmacosx-version-min=10.15 -fPIC -std=gnu++17 -MD -MT Modules/TPC/CMakeFiles/O2QcTPC.dir/src/CheckOfSlices.cxx.o -MF Modules/TPC/CMakeFiles/O2QcTPC.dir/src/CheckOfSlices.cxx.o.d -o Modules/TPC/CMakeFiles/O2QcTPC.dir/src/CheckOfSlices.cxx.o -c /System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/SOURCES/QualityControl/1209/0/Modules/TPC/src/CheckOfSlices.cxx
/System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/SOURCES/QualityControl/1209/0/Modules/TPC/src/CheckOfSlices.cxx:248:56: error: format specifies type 'int' but the argument has type 'float' [-Werror,-Wformat]
      msg->AddText(Form("Outlier, more than %isigma.", mNSigmaBadExpectedPhysicsValue));
                                            ~~         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                            %f
--
      msg->AddText(Form("Outlier, more than %isigma.", mNSigmaBadExpectedPhysicsValue));
                                            ~~         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                            %f
/System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/SOURCES/QualityControl/1209/0/Modules/TPC/src/CheckOfSlices.cxx:250:56: error: format specifies type 'int' but the argument has type 'float' [-Werror,-Wformat]
      msg->AddText(Form("Outlier, more than %isigma.", mNSigmaBadMean));
                                            ~~         ^~~~~~~~~~~~~~
                                            %f
--
      msg->AddText(Form("Outlier, more than %isigma.", mNSigmaBadMean));
                                            ~~         ^~~~~~~~~~~~~~
                                            %f
/System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/SOURCES/QualityControl/1209/0/Modules/TPC/src/CheckOfSlices.cxx:261:56: error: format specifies type 'int' but the argument has type 'float' [-Werror,-Wformat]
      msg->AddText(Form("Outlier, more than %isigma.", mNSigmaExpectedPhysicsValue));
                                            ~~         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
                                            %f
--
      msg->AddText(Form("Outlier, more than %isigma.", mNSigmaExpectedPhysicsValue));
                                            ~~         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
                                            %f
/System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/SOURCES/QualityControl/1209/0/Modules/TPC/src/CheckOfSlices.cxx:263:56: error: format specifies type 'int' but the argument has type 'float' [-Werror,-Wformat]
      msg->AddText(Form("Outlier, more than %isigma.", mNSigmaMean));
                                            ~~         ^~~~~~~~~~~
                  
                  ```

**Comment by wiechula** [comment]
@HorstMa , can you fix the formatting and perhaps also change to `fmt::format`.

**Review by wiechula — CHANGES_REQUESTED** [review]

**Review by wiechula — APPROVED** [review]

**Pull request merged**